README.rst

```diff
@@ -1,102 +1,208 @@
+
+.. _TyDe 2020: http://davidchristiansen.dk/pubs/tyde2020-predictable-macros-abstract.pdf
+.. _video: https://www.youtube.com/watch?v=FyeWwYfqTHo&t=996s
+.. _Compose NYC 2019: https://www.youtube.com/watch?v=nUvKoG_V_U0
+
+::
+
+  #lang "prelude.kl"
+
+  -- do notation is not builtin syntax, it's implemented as a library!
+  (import "monad.kl")
+
+  -- An effectful action whose inferred type is (-> String (IO Unit))
+  (defun putStrLn (str)
+    (write stdout (string-append str "\n")))
+
+  -- "run" is like main, except you can have more than one.
+  (run
+    -- Klister doesn't have typeclasses yet, so "do" needs an explicit
+    -- dictionary argument.
+    (do io-monad
+      (putStrLn "hello")
+      (putStrLn "world")))
+
+You can run the above program using either stack or cabal::
+
+    $ cabal run klister -- run examples/hello.kl
+    hello
+    world
+
+Features
+========
+
+Features we borrow from Racket:
+
+* `Custom syntax`_, via `hygienic macros`_ with `easy-to-override hygiene`_.
+* `Custom languages`_ (``#lang``), via macros which reinterpret terms into
+  those of an existing ``#lang``.
+* `Syntax objects`_, that is, s-expressions annotated with source locations and
+  lexical information.
+* A `module system`_ which allow imports and exports to be renamed.
+
+.. _Custom syntax: examples/lambda-case.golden
+.. _hygienic macros: TODO: write a short example demonstrating lack of capture.
+.. _easy-to-override hygiene: examples/anaphoric-if.kl
+.. _Custom languages: examples/rpn.kl
+.. _Syntax objects: TODO: link to a short example which explains that in
+   Racket, syntax objects are introduced via ``#'(...)``, whereas in Klister
+   they are introduced via ``'(...)``. Also explain that Klister does not have
+   unannotated s-expressions.
+.. _module system: TODO: write a short example demonstrating what can be
+   expressed with the import and export primitives.
+
+Features we borrow from Haskell:
+
+* Statically-typed; we check that macro implementations have type
+  ``(-> Syntax (Macro Syntax)``, and *then* that the generated code is also
+  well-typed. We thus aim for the expressivity of Template Haskell, not the
+  extra guarantees of Typed Template Haskell.
+* Purely functional; primitives with side-effects either run in the ``IO``
+  monad or in the ``Macro`` monad, as appropriate.
+* Higher-kinded types; for example `monads`_ are defined as a library.
+* `Algebraic datatypes`_. In Racket, ``(list 1 2 3)`` and ``'(1 2 3)``
+  construct the same value, an s-expression which satisfies the ``list?``
+  predicate. In Klister, the former produces a value of the algebraic datatype
+  ``(List Integer)``, while the latter produces a value of type ``Syntax``.
+  This syntax can be "opened", via ``(open-syntax '(1 2 3))``, to obtain the
+  value ``(list-contents (list '1 '2 '3))`` of type
+  ``(Syntax-Contents Syntax)``.
+* `Bidirectional type inference`_; the type of an argument can specialize the
```

Comment:
This isn't what's usually meant by bidirectional type inference. We do Hindley-Milner style global type inference, but bidirectional typing usually refers to a particular assemblage of type system features rather than H-M.

Reply:
fixed in 9232b80a847b8c3b51fceeca5bca6c6ab48eeabc